source/adios2/toolkit/burstbuffer/FileDrainer.cpp

```diff
@@ -0,0 +1,279 @@
+        Open(f, path, append);
+        return f;
+    }
+}
+
+void FileDrainer::Open(InputFile f, const std::string &path)
+{
+
+    f->rdbuf()->pubsetbuf(0, 0);
+    f->open(path, std::ios::in);
+}
+
+void FileDrainer::Open(OutputFile f, const std::string &path, bool append)
+{
+
+    if (append)
+    {
+        f->rdbuf()->pubsetbuf(0, 0);
+        f->open(path, std::ios::out | std::ios::app | std::ios::binary);
+    }
+    else
+    {
+        f->rdbuf()->pubsetbuf(0, 0);
+        f->open(path, std::ios::out | std::ios::trunc | std::ios::binary);
+    }
+}
+
+void FileDrainer::Close(InputFile f) { f->close(); }
+void FileDrainer::Close(OutputFile f) { f->close(); }
+
+bool FileDrainer::Good(InputFile f) { return (f->good()); }
+bool FileDrainer::Good(OutputFile f) { return (f->good()); }
+
+void FileDrainer::CloseAll()
+{
+    for (auto it = m_OutputFileMap.begin(); it != m_OutputFileMap.end(); ++it)
+    {
+        if (it->second->good())
+        {
+            Close(it->second);
+        }
+        m_OutputFileMap.erase(it);
+    }
+    for (auto it = m_InputFileMap.begin(); it != m_InputFileMap.end(); ++it)
+    {
+        if (it->second->good())
+        {
+            Close(it->second);
+        }
+        m_InputFileMap.erase(it);
+    }
+}
+
+void FileDrainer::Seek(InputFile f, size_t offset, const std::string &path)
+{
+    f->seekg(offset, std::ios_base::beg);
+}
+
+void FileDrainer::Seek(OutputFile f, size_t offset, const std::string &path)
+{
+    f->seekp(offset, std::ios_base::beg);
+}
+
+void FileDrainer::SeekEnd(OutputFile f) { f->seekp(0, std::ios_base::end); }
+
+std::pair<size_t, double> FileDrainer::Read(InputFile f, size_t count,
+                                            char *buffer,
+                                            const std::string &path)
+{
+    size_t totalRead = 0;
+    double totalSlept = 0.0;
+    const double sleepUnit = 0.01; // seconds
+    while (count > 0)
+    {
+        f->read(buffer, static_cast<std::streamsize>(count));
+        const auto readSize = f->gcount();
+
+        if (readSize < static_cast<std::streamsize>(count))
+        {
+            if (f->eof())
+            {
+                std::chrono::duration<double> d(sleepUnit);
+                std::this_thread::sleep_for(d);
+                f->clear(f->rdstate() & ~std::fstream::eofbit);
+                totalSlept += sleepUnit;
+            }
+            else
+            {
+                throw std::ios_base::failure(
+                    "FileDrainer couldn't read from file " + path +
+                    " count = " + std::to_string(count) + " bytes but only " +
+                    std::to_string(readSize) + "\n");
+                break;
+            }
+        }
+        buffer += readSize;
+        count -= readSize;
+        totalRead += readSize;
+    }
+    return std::pair<size_t, double>(totalRead, totalSlept);
+}
+
+size_t FileDrainer::Write(OutputFile f, size_t count, const char *buffer,
```

Comment:
`const size_t count`  is more clear as `count` is not changing in the function. A local non-const copy to pass it `f->write` would have to be inserted. Unlike POSIX's write fstream is not meant to run in a loop.